# [Helm](https://helm.sh/docs/)

Helm é o gerenciador de pacotes do Kubernetes.

Já vimos como criar cada recurso separadamente. O Helm realiza a gestão dos recursos (ns, deploy, svc, ...) desta forma, com apenas um comando, podemos implantar, alterar e excluir todo nosso projeto.

In [1]:
helm version

version.BuildInfo{Version:"v3.3.1", GitCommit:"249e5215cde0c3fa72e27eb7a30e8d55c9696144", GitTreeState:"clean", GoVersion:"go1.14.7"}


Para criar o helm, executamos o comando abaixo:  

```
helm create \<nome-aplicacao>
```

Este comando irá criar a estrutura abaixo:
```
<nome-aplicacao>/
├── Chart.yaml      # Informações Técnicas
├── values.yaml     # Variáveis e Valores
└── templates/      # Modelos de Deployments, Services e todos outros recursos
```

[Helm do Projeto front-app](https://github.com/kdop-dev/front-app/tree/master/helm/front-app)

## Preparando o acesso ao Ambiente

In [2]:
export KUBECONFIG=$PWD/kubeconfig
mkdir -p work/inovacao
cd $PWD/work/inovacao
echo $PWD

/home/jovyan/work/inovacao


## Projeto front-app

1. Publicando o Projeto front-app

    Para a publicar a aplicação via helm, podemos utilizar duas formas:
    * **helm install** - Comando para publicação da aplicação.
    * **helm upgrade** - Comando para atualização da aplicação. Passando o parâmetro **--install** ele publica a aplicação, caso não exista.

In [4]:
#Informe um nome para o namespace
myNamespace="adsantos"

In [ ]:
#helm install --namespace $myNamespace --create-namespace front-app front-app/helm/front-app
helm upgrade --install --namespace $myNamespace --create-namespace front-app front-app/helm/front-app

2. Listando os helms do namespace

In [6]:
helm list --namespace $myNamespace 

NAME     	NAMESPACE	REVISION	UPDATED                               	STATUS  	CHART          	APP VERSION
back-app 	adsantos 	3       	2020-09-18 17:17:12.239461 -0300 -0300	deployed	back-app-0.1.0 	1.16.0     
cert-app 	adsantos 	2       	2020-09-18 17:36:02.55048 -0300 -0300 	deployed	cert-app-0.1.0 	1.16.0     
front-app	adsantos 	6       	2020-09-18 18:37:55.091191 -0300 -0300	deployed	front-app-0.1.0	1.16.0     


3. Consultando os recursos do namespace

In [7]:
kubectl get all -n $myNamespace

NAME                             READY   STATUS    RESTARTS   AGE
pod/back-app-588746f946-zsppn    1/1     Running   0          3d16h
pod/cert-app-776f86cdf7-rmwbq    1/1     Running   0          3d16h
pod/front-app-7fc6bb89ff-l27w6   1/1     Running   0          3d13h

NAME                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
service/back-app    ClusterIP   10.0.200.175   <none>        5000/TCP   3d16h
service/cert-app    ClusterIP   10.0.254.14    <none>        5000/TCP   3d16h
service/front-app   ClusterIP   10.0.253.1     <none>        5000/TCP   3d16h

NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/back-app    0/1     1            0           3d16h
deployment.apps/cert-app    0/1     1            0           3d16h
deployment.apps/front-app   0/1     1            0           3d16h

NAME                                   DESIRED   CURRENT   READY   AGE
replicaset.apps/back-app-588746f946    1         1         0       3d16h
replica

4. Verificando o problema no POD

In [ ]:
kubectl describe pod <insert-pod-name-here> -n $myNamespace 

* O POD necessita da aplicação back-app para poder ser executada.
  * **livenessProbe**: Verifica a saúde do container. Somente se o container estiver em saudável, o POD entrará em execução.    
```yaml
        livenessProbe:
            httpGet:
              path: /health
              port: http
            initialDelaySeconds: 3 # Tempo para realizar a primeira validação
            periodSeconds: 10      # Tempo em que será realizada a sondagem da atividade.
```
  </br>
  * **readinessProbe**: Indica se o contêiner está pronto para atender às solicitações. Se a análise de prontidão falhar, o controlador de terminais removerá o endereço IP do Pod dos terminais de todos os Serviços que correspondem ao Pod. 
```yaml
        readinessProbe:
            httpGet:
              path: /health
              port: http
            initialDelaySeconds: 3 # Tempo para realizar a primeira validação
            periodSeconds: 10      # Tempo em que será realizada a sondagem da atividade.
```
  </br>
  * **startupProbe**: Indica se o aplicativo no contêiner foi iniciado. Todas as outras análises serão desativadas se uma sonda de inicialização for fornecida, até que seja bem-sucedida. Se a sonda de inicialização falhar, o kubelet matará o Container e o Container será sujeita à sua política de reinicialização.
```yaml
        startupProbe:
            httpGet:
              path: /health
              port: http
            initialDelaySeconds: 3 # Tempo para realizar a primeira validação
            periodSeconds: 10      # Tempo em que será realizada a sondagem da atividade.
```
</br>
    Pode-se verificar a saúde do POD de 3 formas:  </br>
    1. executar um comando dentro de um contêiner   </br>    
    2. fazer uma solicitação HTTP em um contêiner  </br>
    (Código maior ou igual a 200 e menor que 400 indica sucesso. Qualquer outro código indica falha.) </br>
    3. abrir um soquete TCP em um contêiner.


### Como as requisições chegam aos containers?

Existem algumas alternativas para expor nossa aplicação utilizando [serviços](https://kubernetes.io/docs/concepts/services-networking/service/), tais como ClusterIP, LoadBalancer, NodePort, ExternalName. O Ingress não é um serviço é uma configuração para uma aplicação (ingress controller) direcionar o tráfego para o nosso serviço.

![](media/nginx-ingress.png)

Fonte para os ícones: <https://github.com/kubernetes/community/tree/master/icons>

No namespace ingress-nginx está instalado o Ingress Controller, ele é um [Servidor Web Nginx](https://github.com/nginxinc/kubernetes-ingress) modificado e integrado com o kubernetes. Existem outros Ingress Controller, para uma lista mais extensiva acesse [Kubernetes Ingress Controller Overview](https://medium.com/swlh/kubernetes-ingress-controller-overview-81abbaca19ec).

#### Ingress

Como configuramos o nosso ingress? Isso depende de como queremos que ele seja acessado, novamente existem várias opções, as mais comuns são:

* Domínio: front-app.com
* Subdominio: front-app.kdop.net
* Caminho: kdop.net/front-app

Para esta aplicação configuramos a terceira opção. Vamos abrir o arquivo [values.yaml](work/inovacao/front-app/helm/front-app/values.yaml) para entender como foi configurado.

```yaml
service:
  type: ClusterIP
  port: 5000

ingress:
  enabled: true
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /$2
  hosts:
    - host: learn.kdop.net
      paths:
      - /adsantos/front-app(/|$)(.*)
  tls: []
```

Esses parâmetros, quando aplicados pelo helm aos templates de serviço e ingress irão criar a seguinte configuração:

```yaml
---
# Source: front-app/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: front-app
  labels:
    helm.sh/chart: front-app-0.1.0
    app.kubernetes.io/name: front-app
    app.kubernetes.io/instance: front-app
    app.kubernetes.io/version: "1.16.0"
    app.kubernetes.io/managed-by: Helm
spec:
  type: ClusterIP
  ports:
    - port: 5000
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: front-app
    app.kubernetes.io/instance: front-app
---
# Source: front-app/templates/ingress.yaml
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: front-app
  labels:
    helm.sh/chart: front-app-0.1.0
    app.kubernetes.io/name: front-app
    app.kubernetes.io/instance: front-app
    app.kubernetes.io/version: "1.16.0"
    app.kubernetes.io/managed-by: Helm
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /$2
spec:
  rules:
    - host: "learn.kdop.net"
      http:
        paths:
          - path: /adsantos/front-app(/|$)(.*)
            backend:
              serviceName: front-app
              servicePort: 5000
```

> Você pode imprimir o que será criado pelo helm com o comando `helm template` no lugar de `helm install`.

> Sobre TLS ver [README-ingresss-with-tls.md](README-ingresss-with-tls.md).

Um serviço do tipo **ClusterIP** foi criado e seu seletor são os PODs com labels `app.kubernetes.io/name` e `app.kubernetes.io/instance` com o valor `front-app`. Não importa quantas réplicas existirem, o serviço irá direcionar o tráfego, na forma de _round-robin_, para cada uma delas.

Mas _ClusterIP_ é um tipo de serviço que só é visível dentro do cluster, pode ser referênciado pelos containers através do seu nome e porta, neste caso `ping front-app` resultaria no retorno do IP associado ao serviço e `telnet front-app 5000` em uma conexão de sucesso.

Para acessar os containers de fora do cluster foi criado uma configuração tipo Ingress, nela foi definido a configuração do frontend que deverá direcionar o tráfego para o backend. Neste cenário, a URL que irá direcionar o tráfego para os nossos containers é http://learn.kdop.net/adsantos/front-app e qualquer caminhos e parâmetros adicionais serão repassados para os nossos containers na forma http://front-app:5000/<o resto da url>.

In [10]:
kubectl describe ingress -n $myNamespace

Name:             cert-app
Namespace:        adsantos
Address:          13.86.33.72
Default backend:  default-http-backend:80 (<error: endpoints "default-http-backend" not found>)
Rules:
  Host                          Path  Backends
  ----                          ----  --------
  learn.kdop.net                
                                /adsantos/cert-app(/|$)(.*)   cert-app:5000 ()
  kdop-learn.krthomolog.com.br  
                                /adsantos/cert-app(/|$)(.*)   cert-app:5000 ()
Annotations:                    meta.helm.sh/release-name: cert-app
                                meta.helm.sh/release-namespace: adsantos
                                nginx.ingress.kubernetes.io/rewrite-target: /$2
Events:                         <none>


Name:             front-app
Namespace:        adsantos
Address:          13.86.33.72
Default backend:  default-http-backend:80 (<error: endpoints "default-http-backend" not found>)
Rules:
  Host                          Path  Backend

Mas para essa configuração funcionar é necessário que um Ingress-Controller esteja instalado e lendo o seu namespace a procura desse tipo de configuração.
    
Neste cluster, o Ingress-Controller instalado é o Nginx, um servidor de web, que funciona como proxy, com um LoadBalancer externo direcionando o tráfego.

In [9]:
kubectl get all -n ingress-nginx

NAME                                            READY   STATUS      RESTARTS   AGE
pod/ingress-nginx-admission-create-85qfk        0/1     Completed   0          9d
pod/ingress-nginx-admission-patch-vf2wd         0/1     Completed   0          9d
pod/ingress-nginx-controller-5859f5cd59-9gn6t   1/1     Running     0          3d13h

NAME                                         TYPE           CLUSTER-IP     EXTERNAL-IP   PORT(S)                      AGE
service/ingress-nginx-controller             LoadBalancer   10.0.99.196    13.86.33.72   80:30238/TCP,443:31738/TCP   9d
service/ingress-nginx-controller-admission   ClusterIP      10.0.156.183   <none>        443/TCP                      9d

NAME                                       READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/ingress-nginx-controller   0/1     1            0           9d

NAME                                                  DESIRED   CURRENT   READY   AGE
replicaset.apps/ingress-nginx-controller-5859f5cd59   1 

O serviço `ingress-nginx-controller` é do tipo **LoadBalancer** e seu endereço externo foi configurado em um DNS (AWS Route 53) para apontar para o seu IP.

In [8]:
kubectl describe service/ingress-nginx-controller -n ingress-nginx

Name:                     ingress-nginx-controller
Namespace:                ingress-nginx
Labels:                   app.kubernetes.io/component=controller
                          app.kubernetes.io/instance=ingress-nginx
                          app.kubernetes.io/managed-by=Helm
                          app.kubernetes.io/name=ingress-nginx
                          app.kubernetes.io/version=0.35.0
                          helm.sh/chart=ingress-nginx-2.13.0
Annotations:              <none>
Selector:                 app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx,app.kubernetes.io/name=ingress-nginx
Type:                     LoadBalancer
IP:                       10.0.99.196
LoadBalancer Ingress:     13.86.33.72
Port:                     http  80/TCP
TargetPort:               http/TCP
NodePort:                 http  30238/TCP
Endpoints:                
Port:                     https  443/TCP
TargetPort:               https/TCP
NodePort:              

Aqui você pode obter o IP do Ingress-Controller e a quais portas ele esta associado, neste caso 80 (http) e 443 (https).

[<img align="left" src="media/voltar.png" width="100"/>](02_Kubernetes.ipynb) [<img align="right" src="media/avancar.png" width="100"/>](04_Back.ipynb)
